### PR TITLE
Cranelift upgrade

### PR DESCRIPTION
**Pull request by syrusakbary**
This PR updates Cranelift to it's latest version `0.101.4`.

### CODE REVIEW

File: lib/compiler-cranelift/src/heap.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+/// generate a trap when accessed.
+///
+/// #### Dynamic heaps
+///
+/// A *dynamic heap* can be relocated to a different base address when it is
+/// resized, and its bound can move dynamically. The offset-guard pages move
+/// when the heap is resized. The bound of a dynamic heap is stored in a global
+/// value.
+#[derive(Clone, PartialEq, Hash)]
+#[cfg_attr(
```
Comment (theduke) [inline lib/compiler-cranelift/src/heap.rs:65]: Did we copy this from somewhere?
Does it need attribution?

File: lib/compiler-cranelift/src/translator/code_translator/bounds_checks.rs [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+//!
+//! Given
+//!
+//! * a dynamic Wasm memory index operand,
+//! * a static offset immediate, and
+//! * a static access size,
+//!
+//! bounds check the memory access and translate it into a native memory access.
+//!
+//! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (theduke) [inline lib/compiler-cranelift/src/translator/code_translator/bounds_checks.rs:11]: As above, is this copied?

File: lib/compiler-cranelift/src/heap.rs (hunk 1 above)
Comment (syrusakbary) [reply lib/compiler-cranelift/src/heap.rs:65]: Yup, is copied but already attributed:
https://github.com/wasmerio/wasmer/tree/6df1e38bd9af4fba2219cf1e54aba975f0f39b77/lib/compiler-cranelift#acknowledgments

File: lib/compiler-cranelift/src/translator/code_translator/bounds_checks.rs (hunk 2 above)
Comment (syrusakbary) [reply lib/compiler-cranelift/src/translator/code_translator/bounds_checks.rs:11]: Yup, is copied but already attributed:
https://github.com/wasmerio/wasmer/tree/6df1e38bd9af4fba2219cf1e54aba975f0f39b77/lib/compiler-cranelift#acknowledgments

File: lib/compiler-cranelift/src/heap.rs (hunk 1 above)
Comment (theduke) [reply lib/compiler-cranelift/src/heap.rs:65]: We should put that in the files so future readers know where it comes from (including us)

**Comment by xdoardo** [comment]
Closing in favour of #5019.

**Pull request closed without merging**